README.md
# SGQR Parser

Parse SGQR code (Singapore Quick Response Code) as per http://www.mas.gov.sg/sgqr.
No validation is performed at this point.

Gave a talk on this at the December 2018 meetup for the Singapore PHP User Group - video recording
[here](https://engineers.sg/video/writing-a-sgqr-parser-singapore-php-user-group--3061).

## Requirements
- PHP >= 7.0
- Composer >= 1.6.4

## Installation
- Clone this repo.
- Run `composer install` to install dependencies.

## Tests
- Run `composer test` to run tests.
- Run `composer cs` to run coding style checks.
- Run `composer check` to do both.

## Add to project
- As this library is not available on Packagist, the GitHub repository needs to be added to the
  project's `composer.json`.

  ```
  {
    "repositories": [
        {
            "type": "git",
            "url": "https://github.com/zionsg/sgqr-parser"
        }
    ]
  }
  ```
- Run `composer require zionsg/sgqr-parser:dev-master` to get the latest release in the `master` branch.

## Usage
- See PHPUnit tests in `test` folder.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file, in reverse chronological order by release.
The format follows [Keep a CHANGELOG](http://keepachangelog.com/).
This project adheres to [Semantic Versioning](http://semver.org/).

## [v0.4.1] - 2021-03-10

### Added
- Nothing.

### Changed
- Nothing.

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- `src/SgQr/Parser.php:computeChecksum()`: Pad return value to 4 characters.

### Security
- Nothing.

## [v0.4.0] - 2020-12-19

### Added
- `src/SgQr/Parser.php`:
    + `computeChecksum()`: Added method to compute checksum for SGQR code.
    + `assemble())`: Added method to assemble SGQR.

### Changed
- Nothing.

### Deprecated
- Nothing.

### Removed
- Nothing.

### Fixed
- Nothing.

### Security
- Nothing.

## [v0.3.0] - 2018-10-31

### Added
- Handling of nested templates, in particular data objects ID 50 - 99 under the template for root data object ID 62.
- Mark nested templates with `isTemplate` key in specifications.
- Coding style checks using PHP CodeSniffer.

### Changed
- Use PHPUnit for tests.

### Removed
- Nothing.

### Fixed
- Nothing.

## [v0.2.0] - 2018-10-22

### Added
- This CHANGELOG file.

### Changed
- Flatten `id` array in output. Difference shown below for a single data object in the output:

  ```
  // Previously
  [
      'id' => [
          'key' => '00',
          'name' => 'Payload Format Indicator',
      ],
      'length' => '02',
      'value' => '01',
      'comment' => 'Shall be the 1st data object in QR code. Shall contain value of 01.',
  ]

  // Now
  [
      'id' => '00',
      'name' => 'Payload Format Indicator',
      'length' => '02',
      'value' => '01',
      'comment' => 'Shall be the 1st data object in QR code. Shall contain value of 01.',
  ]
  ```

### Removed
- Nothing.

### Fixed
- Nothing.

## [v0.1.0] - 2018-10-21

1st release. Used at https://intzone.com/tools/sgqr.

[Unreleased]: https://github.com/zionsg/sgqr-parser/compare/v0.4.1...HEAD
[v0.4.1]: https://github.com/zionsg/sgqr-parser/compare/v0.4.0...v0.4.1
[v0.4.0]: https://github.com/zionsg/sgqr-parser/compare/v0.3.0...v0.4.0
[v0.3.0]: https://github.com/zionsg/sgqr-parser/compare/v0.2.0...v0.3.0
[v0.2.0]: https://github.com/zionsg/sgqr-parser/compare/v0.1.0...v0.2.0
[v0.1.0]: https://github.com/zionsg/sgqr-parser/tree/v0.1.0


tests/test.php
<?php
$dir = __DIR__ . '/../';
require $dir . 'vendor/autoload.php';

// SGQR code sample from SOBA EXPRESS in article https://www.channelnewsasia.com/news/singapore/sgqr-qr-code-cashless-payment-singapore-rolls-out-unified-10727568
$qrCode = '00020101021126810011SG.COM.NETS01231198500065G9912312359000211111686614000308686614019908604108C251800007SG.SGQR01121809072DD85C020701.00010306079027040201050206060400000708201809155204581253037025802SG5912SOBA EXPRESS6009Singapore630457B3';

$parser = new \SgQr\Parser();
$result = $parser->parse($qrCode);
echo json_encode($result, JSON_PRETTY_PRINT);


test/SgQr/ParserTest.php
<?php
namespace SgQrTest;

use SgQr\Parser;
use PHPUnit\Framework\TestCase;

final class ParserTest extends TestCase
{
    /**
     * @dataProvider qrCodeProvider
     */
    public function testCanParse($dataSetName, $qrCode, $expectedResultInJson)
    {
        $parser = new Parser();
        $result = json_encode($parser->parse($qrCode));

        $this->assertEquals($expectedResultInJson, $result);
    }

    /**
     * @dataProvider qrCodeProvider
     */
    public function testChecksum($dataSetName, $qrCode, $expectedResultInJson)
    {
        $parser = new Parser();
        $expectedChecksum = substr($qrCode, -4);
        $actualChecksum = $parser->computeChecksum(substr($qrCode, 0, -4));

        $this->assertEquals($expectedChecksum, $actualChecksum);
    }

    /**
     * @dataProvider qrCodeProvider
     */
    public function testCanAssemble($dataSetName, $qrCode, $expectedResultInJson)
    {
        $parser = new Parser();
        $parsed = json_decode(trim($expectedResultInJson), true);
        $assembled = $parser->assemble($parsed);

        $this->assertEquals($qrCode, $assembled);
    }

    public function qrCodeProvider()
    {
        return [
            [   // SGQR for SOBA EXPRESS from article https://www.channelnewsasia.com/news/singapore/sgqr-qr-code-cashless-payment-singapore-rolls-out-unified-10727568
                'Basic SGQR',
                '00020101021126810011SG.COM.NETS01231198500065G9912312359000211111686614000308686614019908604108C251800007SG.SGQR01121809072DD85C020701.00010306079027040201050206060400000708201809155204581253037025802SG5912SOBA EXPRESS6009Singapore630457B3',
                '[{"id":"00","name":"Payload Format Indicator","length":"02","value":"01","comment":"Shall be the 1st data object in QR code. Shall contain value of 01."},{"id":"01","name":"Point of Initiation Method","length":"02","value":"11","comment":"Value of 11 used when same QR code used for more than 1 transaction. Value of 12 used when a new QR code is shown for each transaction."},{"id":"26","name":"Merchant Account Information","length":"81","value":"0011SG.COM.NETS01231198500065G9912312359000211111686614000308686614019908604108C2","comment":"Templates reserved for additional payment networks.","dataObjects":[{"id":"00","name":"Globally Unique Identifier","length":"11","value":"SG.COM.NETS","comment":"Value can be one of the following - an Application Identifier (AID), a UUID without hyphens, a reverse domain name."},{"id":"01","name":"Payment network specific","length":"23","value":"1198500065G991231235900","comment":""},{"id":"02","name":"Payment network specific","length":"11","value":"11168661400","comment":""},{"id":"03","name":"Payment network specific","length":"08","value":"68661401","comment":""},{"id":"99","name":"Payment network specific","length":"08","value":"604108C2","comment":""}]},{"id":"51","name":"Merchant Account Information","length":"80","value":"0007SG.SGQR01121809072DD85C020701.0001030607902704020105020606040000070820180915","comment":"Templates reserved for additional payment networks.","dataObjects":[{"id":"00","name":"Globally Unique Identifier","length":"07","value":"SG.SGQR","comment":"Value can be one of the following - an Application Identifier (AID), a UUID without hyphens, a reverse domain name."},{"id":"01","name":"Payment network specific","length":"12","value":"1809072DD85C","comment":""},{"id":"02","name":"Payment network specific","length":"07","value":"01.0001","comment":""},{"id":"03","name":"Payment network specific","length":"06","value":"079027","comment":""},{"id":"04","name":"Payment network specific","length":"02","value":"01","comment":""},{"id":"05","name":"Payment network specific","length":"02","value":"06","comment":""},{"id":"06","name":"Payment network specific","length":"04","value":"0000","comment":""},{"id":"07","name":"Payment network specific","length":"08","value":"20180915","comment":""}]},{"id":"52","name":"Merchant Category Code","length":"04","value":"5812","comment":"As defined by ISO 18245."},{"id":"53","name":"Transaction Currency","length":"03","value":"702","comment":"3-digit numeric representation of currency according to ISO 4217. USD is 840, SGD is 702."},{"id":"58","name":"Country Code","length":"02","value":"SG","comment":"As defined by ISO 3166-1 alpha 2."},{"id":"59","name":"Merchant Name","length":"12","value":"SOBA EXPRESS","comment":""},{"id":"60","name":"Merchant City","length":"09","value":"Singapore","comment":""},{"id":"63","name":"CRC","length":"04","value":"57B3","comment":"Shall be the last data object in QR code. Checksum calculated according to ISO\/IEC 13239 using polynomial 1021 (hex) and initial value FFFF (hex)."}]',
            ],

            [   // SGQR from https://www.nets.com.sg/promotions/presidents-star-charity-2018/
                'SGQR contains single quote',
                "00020101021126810011SG.COM.NETS01231198500065G9912312359000211111895264000308895264019908735A43CF51800007SG.SGQR01121809172E1FE7020701.00010306279621040201050202060400000708201809175204000053037025802SG5925PRESIDENT'S CHALLENGE NO16009Singapore63048DCC",
                '[{"id":"00","name":"Payload Format Indicator","length":"02","value":"01","comment":"Shall be the 1st data object in QR code. Shall contain value of 01."},{"id":"01","name":"Point of Initiation Method","length":"02","value":"11","comment":"Value of 11 used when same QR code used for more than 1 transaction. Value of 12 used when a new QR code is shown for each transaction."},{"id":"26","name":"Merchant Account Information","length":"81","value":"0011SG.COM.NETS01231198500065G9912312359000211111895264000308895264019908735A43CF","comment":"Templates reserved for additional payment networks.","dataObjects":[{"id":"00","name":"Globally Unique Identifier","length":"11","value":"SG.COM.NETS","comment":"Value can be one of the following - an Application Identifier (AID), a UUID without hyphens, a reverse domain name."},{"id":"01","name":"Payment network specific","length":"23","value":"1198500065G991231235900","comment":""},{"id":"02","name":"Payment network specific","length":"11","value":"11189526400","comment":""},{"id":"03","name":"Payment network specific","length":"08","value":"89526401","comment":""},{"id":"99","name":"Payment network specific","length":"08","value":"735A43CF","comment":""}]},{"id":"51","name":"Merchant Account Information","length":"80","value":"0007SG.SGQR01121809172E1FE7020701.0001030627962104020105020206040000070820180917","comment":"Templates reserved for additional payment networks.","dataObjects":[{"id":"00","name":"Globally Unique Identifier","length":"07","value":"SG.SGQR","comment":"Value can be one of the following - an Application Identifier (AID), a UUID without hyphens, a reverse domain name."},{"id":"01","name":"Payment network specific","length":"12","value":"1809172E1FE7","comment":""},{"id":"02","name":"Payment network specific","length":"07","value":"01.0001","comment":""},{"id":"03","name":"Payment network specific","length":"06","value":"279621","comment":""},{"id":"04","name":"Payment network specific","length":"02","value":"01","comment":""},{"id":"05","name":"Payment network specific","length":"02","value":"02","comment":""},{"id":"06","name":"Payment network specific","length":"04","value":"0000","comment":""},{"id":"07","name":"Payment network specific","length":"08","value":"20180917","comment":""}]},{"id":"52","name":"Merchant Category Code","length":"04","value":"0000","comment":"As defined by ISO 18245."},{"id":"53","name":"Transaction Currency","length":"03","value":"702","comment":"3-digit numeric representation of currency according to ISO 4217. USD is 840, SGD is 702."},{"id":"58","name":"Country Code","length":"02","value":"SG","comment":"As defined by ISO 3166-1 alpha 2."},{"id":"59","name":"Merchant Name","length":"25","value":"PRESIDENT\'S CHALLENGE NO1","comment":""},{"id":"60","name":"Merchant City","length":"09","value":"Singapore","comment":""},{"id":"63","name":"CRC","length":"04","value":"8DCC","comment":"Shall be the last data object in QR code. Checksum calculated according to ISO\/IEC 13239 using polynomial 1021 (hex) and initial value FFFF (hex)."}]',
            ],

            [   // Modified SGQR generated for PayNow using POSB banking app, with valid CRC checksum
                'Test nested template for root data object ID 62',
                '00020101021126500009SG.PAYNOW010100211+65123456780301004082030123152040000530370254071234.565802SG5902ME6009Singapore62290106BILLNO50150004TEST0103ABC6304B301',
                '[{"id":"00","name":"Payload Format Indicator","length":"02","value":"01","comment":"Shall be the 1st data object in QR code. Shall contain value of 01."},{"id":"01","name":"Point of Initiation Method","length":"02","value":"11","comment":"Value of 11 used when same QR code used for more than 1 transaction. Value of 12 used when a new QR code is shown for each transaction."},{"id":"26","name":"Merchant Account Information","length":"50","value":"0009SG.PAYNOW010100211+651234567803010040820301231","comment":"Templates reserved for additional payment networks.","dataObjects":[{"id":"00","name":"Globally Unique Identifier","length":"09","value":"SG.PAYNOW","comment":"Value can be one of the following - an Application Identifier (AID), a UUID without hyphens, a reverse domain name."},{"id":"01","name":"Payment network specific","length":"01","value":"0","comment":""},{"id":"02","name":"Payment network specific","length":"11","value":"+6512345678","comment":""},{"id":"03","name":"Payment network specific","length":"01","value":"0","comment":""},{"id":"04","name":"Payment network specific","length":"08","value":"20301231","comment":""}]},{"id":"52","name":"Merchant Category Code","length":"04","value":"0000","comment":"As defined by ISO 18245."},{"id":"53","name":"Transaction Currency","length":"03","value":"702","comment":"3-digit numeric representation of currency according to ISO 4217. USD is 840, SGD is 702."},{"id":"54","name":"Transaction Amount","length":"07","value":"1234.56","comment":"Shall not be included if the app should prompt user to enter the amount to be paid to the Merchant."},{"id":"58","name":"Country Code","length":"02","value":"SG","comment":"As defined by ISO 3166-1 alpha 2."},{"id":"59","name":"Merchant Name","length":"02","value":"ME","comment":""},{"id":"60","name":"Merchant City","length":"09","value":"Singapore","comment":""},{"id":"62","name":"Additional Data Field Template","length":"29","value":"0106BILLNO50150004TEST0103ABC","comment":"","dataObjects":[{"id":"01","name":"Bill Number","length":"06","value":"BILLNO","comment":""},{"id":"50","name":"Payment System specific templates","length":"15","value":"0004TEST0103ABC","comment":"","dataObjects":[{"id":"00","name":"Globally Unique Identifier","length":"04","value":"TEST","comment":"Value can be one of the following - an Application Identifier (AID), a UUID without hyphens, a reverse domain name."},{"id":"01","name":"Payment network specific","length":"03","value":"ABC","comment":""}]}]},{"id":"63","name":"CRC","length":"04","value":"B301","comment":"Shall be the last data object in QR code. Checksum calculated according to ISO\/IEC 13239 using polynomial 1021 (hex) and initial value FFFF (hex)."}]',
            ],
        ];
    }
}


src/App/SgQrParser.php
<?php
namespace App;

class SgQrParser
{
    const ROOT_DATA_OBJECTS_BY_ID = 'rootDataObjectsById';
    const INFO_TEMPLATE = 'infoTemplate';
    const TEMPLATES_BY_ID = 'templatesById';
    const NAME = 'name';
    const COMMENT = 'comment';
    const DATA_OBJECTS_BY_ID = 'dataObjectsById';
    const IS_TEMPLATE = 'isTemplate';
    const USES_INFO_TEMPLATE = 'usesInfoTemplate';

    /** @var array */
    protected $specs;

    /** @var array */
    protected $infoTemplate;

    public function __construct(array $specs)
    {
        $this->specs = $specs;
        $this->infoTemplate = $specs[self::INFO_TEMPLATE] ?? [];
    }

    public function parse($qrCode)
    {
        return $this->extractDataObjects($qrCode);
    }

    protected function extractDataObjects($text, $isTemplate = false, $rootId = null)
    {
        $result = [];

        $textLen = strlen($text);
        $index = 0;
        while ($index < $textLen) {
            $id = substr($text, $index, 2);
            $length = substr($text, $index + 2, 2);
            $value = substr($text, $index + 4, $length);
            $index += 4 + $length;

            $result[] = $this->analyzeDataObject($id, $length, $value, $isTemplate, $rootId);
        }

        return $result;
    }

    protected function analyzeDataObject(
        $id,
        $length,
        $value,
        $isTemplate = false,
        $rootId = null
    ) {
        $result = [];

        $info = [];
        if ($isTemplate) {
            $templateInfo = $this->specs[self::TEMPLATES_BY_ID][$rootId] ?? [];
            if ($templateInfo[self::USES_INFO_TEMPLATE] ?? false) {
                $templateInfo[self::DATA_OBJECTS_BY_ID] = $this->infoTemplate;
            }

            $info = $templateInfo[self::DATA_OBJECTS_BY_ID][$id] ?? [];
        } else {
            $info = $this->specs[self::ROOT_DATA_OBJECTS_BY_ID][$id] ?? [];
        }

        if ($info[self::USES_INFO_TEMPLATE] ?? false) {
            $info[self::DATA_OBJECTS_BY_ID] = $this->infoTemplate;
        }

        // Resolve data object
        $result['id'] = [
            'key' => $id,
            'name' => $info[self::NAME] ?? '',
        ];
        $result['length'] = $length;
        $result['value'] = $value;
        $result['comment'] = $info[self::COMMENT] ?? '';

        // Parse data object further if it is a template
        $isObjTemplate = $info[self::IS_TEMPLATE] ?? false;
        if ($isObjTemplate) {
            $result['dataObjects'] = $this->extractDataObjects($value, $isObjTemplate, $id);
        }

        return $result;
    }
}


src/SgQr/Parser.php
<?php
namespace SgQr;

use InvalidArgumentException;

class Parser
{
    const ROOT_DATA_OBJECTS_BY_ID = 'rootDataObjectsById';
    const INFO_TEMPLATE = 'infoTemplate';
    const TEMPLATES_BY_ID = 'templatesById';
    const ID = 'id';
    const NAME = 'name';
    const VALUE = 'value';
    const COMMENT = 'comment';
    const MAX_VALUE_LENGTH = 99;
    const DATA_OBJECTS_BY_ID = 'dataObjectsById';
    const IS_TEMPLATE = 'isTemplate';
    const USES_INFO_TEMPLATE = 'usesInfoTemplate';
    const DATA_OBJECTS = 'dataObjects';

    /** @var array */
    protected $specs;

    /** @var array */
    protected $infoTemplate;

    /**
     * Constructor
     *
     * @param array $specs SGQR specifications. If not specified, default specs will be loaded.
     */
    public function __construct(array $specs = null)
    {
        $specsPath = __DIR__ . '/../../config/sgqr-specs.php';
        if (null === $specs && file_exists($specsPath)) {
            $specs = include $specsPath;
            $specs = is_array($specs) ? $specs : [];
        }

        $this->specs = $specs;
        $this->infoTemplate = $specs[self::INFO_TEMPLATE] ?? [];
    }

    /**
     * Parse SGQR
     *
     * @param string $qrCode Contents of SGQR
     * @return array
     */
    public function parse($qrCode)
    {
        return $this->extractDataObjects($qrCode);
    }

    /**
     * Assemble SGQR
     *
     * This is the opposite of parse().
     *
     * @param array $dataObjects Extracted data objects as per return value of parse() including checksum
     * @param boolean $isRecursiveCall Default=false. Used internally within method, not meant for user to specify.
     * @return string Checksum will be recalculated
     * @throws InvalidArgumentException If any data object is invalid
     */
    public function assemble($dataObjects, $isRecursiveCall = false)
    {
        $result = '';

        foreach (($dataObjects ?: []) as $dataObject) {
            $id = $dataObject[self::ID] ?? '';
            $value = $dataObject[self::VALUE] ?? '';
            $subDataObjects = $dataObject[self::DATA_OBJECTS] ?? [];
            if (!$id || ('' === $value && !$subDataObjects)) {
                throw new InvalidArgumentException('Invalid data object: ' . json_encode($dataObject));
            }

            // Recompute value from inner data objects
            if ($subDataObjects) {
                $value = $this->assemble($subDataObjects, true);
            }

            // Recompute length
            $length = strlen($value);
            if ($length > self::MAX_VALUE_LENGTH) {
                throw new InvalidArgumentException(
                    'Value cannot be more than ' . self::MAX_VALUE_LENGTH . ' characters: '
                    . json_encode($dataObject)
                );
            }

            // Assemble section
            $section = $id . str_pad($length, 2, '0', STR_PAD_LEFT) . $value;
            if (!$section) {
                throw InvalidArgumentException('Could not assemble data object: ' . json_encode($dataObject));
            }

            // Append section
            $result .= $section;
        }

        // Recompute checksum only for original method call, not recursive calls
        if (!$isRecursiveCall) {
            $resultWithoutChecksum = substr($result, 0, -4);
            $result = $resultWithoutChecksum . $this->computeChecksum($resultWithoutChecksum);
        }

        return $result;
    }

    /**
     * Compute CRC-CCITT checksum according to ISO/IEC 13239
     *
     * This yields the same checksum as SGQR - pass the contents of a SGQR code minus the last 4 characters to this
     * function and the result should be the same as the last 4 characters of the contents of the SGQR code.
     *
     * @link Online CRC calculator at https://www.tahapaksu.com/crc/
     * @link From https://stackoverflow.com/questions/30035582/how-to-calculate-crc16-ccitt-in-php-hex/30036901#30036901
     * @param string $text Plain text to compute checksum for, e.g. "Hello World!".
     * @param int $polynomialHex Default=0x1021. Polynomial (in hex) to use.
     * @param int $initialValue Default=0xFFFF. Initial value (in hex) to use.
     * @return string 4-digit hex string, e.g. 882A for text "Hello World!".
     */
    public function computeChecksum($text, $polynomialHex = 0x1021, $initialValue = 0xFFFF)
    {
        $result = $initialValue;
        $buffer = $text;
        $length = strlen($buffer);

        if ($length > 0) {
            for ($offset = 0; $offset < $length; $offset++) {
                $result ^= (ord($buffer[$offset]) << 8);
                for ($bitwise = 0; $bitwise < 8; $bitwise++) {
                    if (($result <<= 1) & 0x10000) {
                        $result ^= $polynomialHex;
                    }

                    $result &= 0xFFFF; // not sure if it is always 0xFFFF or should be $initialValue
                }
            }
        }

        return str_pad(strtoupper(dechex($result)), 4, '0', STR_PAD_LEFT); // ensure 4 chars cos result may be shorter
    }

    /**
     * Extract data objects from SQQR
     *
     * @param string $text Contents of SGQR
     * @param array $specs Specifications for data objects under parent
     * @return array
     */
    protected function extractDataObjects($text, array $specs = null)
    {
        $result = [];

        if (null === $specs) {
            $specs = $this->specs[self::ROOT_DATA_OBJECTS_BY_ID];
        }

        $textLen = strlen($text);
        $index = 0;
        while ($index < $textLen) {
            $id = substr($text, $index, 2);
            $length = substr($text, $index + 2, 2);
            $value = substr($text, $index + 4, $length);
            $index += 4 + $length;

            $result[] = $this->analyzeDataObject($id, $length, $value, $specs[$id] ?? []);
        }

        return $result;
    }

    /**
     * Analyze single data object
     *
     * @param string $id ID of data object
     * @param int $length Length of value for data object
     * @param string $value Value for data object
     * @param array $specs Specifications for data object
     * @return array
     */
    protected function analyzeDataObject($id, $length, $value, array $specs)
    {
        // Does specs belong to a template? If yes, get template info and child data objects if any.
        $isTemplate = $specs[self::IS_TEMPLATE] ?? false;
        if ($isTemplate) {
            $specs = array_merge($specs, $this->specs[self::TEMPLATES_BY_ID][$id] ?? []);
        }

        // Does specs use info template? If yes, populate child data objects with it.
        $usesInfoTemplate = $specs[self::USES_INFO_TEMPLATE] ?? false;
        if ($usesInfoTemplate) {
            $specs[self::DATA_OBJECTS_BY_ID] = $this->infoTemplate;
        }

        // Resolve data object
        $result = [
            'id' => $id,
            'name' => $specs[self::NAME] ?? '',
            'length' => $length,
            'value' => $value,
            'comment' => $specs[self::COMMENT] ?? '',
        ];

        // Parse data object further if it is a template
        if ($isTemplate) {
            $result[self::DATA_OBJECTS] = $this->extractDataObjects($value, $specs[self::DATA_OBJECTS_BY_ID] ?? []);
        }

        return $result;
    }
}
